matrix/src/extension.rs

```diff
@@ -0,0 +1,103 @@
+use core::iter;
+use core::marker::PhantomData;
+
+use p3_field::{ExtensionField, Field};
+
+use crate::Matrix;
+
+#[derive(Debug)]
+pub struct FlatMatrixView<F, EF, Inner>(Inner, PhantomData<(F, EF)>);
+
+impl<F, EF, Inner> FlatMatrixView<F, EF, Inner> {
+    pub fn new(inner: Inner) -> Self {
+        Self(inner, PhantomData)
+    }
+    pub fn inner_ref(&self) -> &Inner {
+        &self.0
+    }
+}
+
+impl<F, EF, Inner> Matrix<F> for FlatMatrixView<F, EF, Inner>
+where
+    F: Field,
+    EF: ExtensionField<F>,
+    Inner: Matrix<EF>,
+{
+    fn width(&self) -> usize {
+        self.0.width() * EF::D
+    }
+
+    fn height(&self) -> usize {
+        self.0.height()
+    }
+
+    type Row<'a> = FlatIter<F, Inner::Row<'a>>
+    where
+        Self: 'a;
+
+    fn row(&self, r: usize) -> Self::Row<'_> {
+        FlatIter {
```

Comment:
I wonder if we could later change `row` to return `impl Iterator<...>`, so implementations like this could leverage `flat_map`?

Reply:
So I tried to hack it this way for a while, but it eventually gets annoying not being able to name the Row type, especially for Send + Sync bounds. You end up with nested impl traits and HRTBs for things like par_rows that percolate downstream